seqr/views/utils/terra_api_utils.py

```diff
@@ -0,0 +1,157 @@
+# This module provides python bindings for the AnVIL Terra API.
+
+import json
+import logging
+import time
+import requests
+
+from urllib.parse import urljoin
+
+from django.core.exceptions import PermissionDenied
+from social_django.models import UserSocialAuth
+from social_django.utils import load_strategy
+
+from settings import SEQR_VERSION, TERRA_API_ROOT_URL
+
+SEQR_USER_AGENT = "seqr/" + SEQR_VERSION
+
+logger = logging.getLogger(__name__)
+
+
+class TerraAPIException(Exception):
+    pass
+
+
+class TerraNotFoundException(TerraAPIException):
+    pass
+
+
+def anvil_enabled():
+    return bool(TERRA_API_ROOT_URL)
+
+
+def is_google_authenticated(user):
+    if not anvil_enabled():
+        return False
+    try:
+        _ = user.social_auth.get(provider = 'google-oauth2')
+    except UserSocialAuth.DoesNotExist:  # Exception happen when the user has never logged-in with Google
+        return False
+    return True
+
+
+def _get_call_args(path, headers=None, root_url=None):
+    if headers is None:
+        headers = {"User-Agent": SEQR_USER_AGENT}
+    headers.update({"accept": "application/json"})
+    if root_url is None:
+        root_url = TERRA_API_ROOT_URL
+    url = urljoin(root_url, path)
+    return url, headers
+
+
+def _get_social_access_token(user):
+    social = user.social_auth.get(provider = 'google-oauth2')
+    if (social.extra_data['auth_time'] + social.extra_data['expires'] - 10) <= int(
+            time.time()):  # token expired or expiring?
+        strategy = load_strategy()
+        try:
+            social.refresh_token(strategy)
+        except Exception as ee:
+            logger.warning('Refresh token failed. {}'.format(str(ee)))
+            raise TerraAPIException('Refresh token failed. {}'.format(str(ee)))
+    return social.extra_data['access_token']
+
+
+def anvil_call(method, path, access_token, user=None, headers=None, root_url=None, **kwargs):
+    url, headers = _get_call_args(path, headers, root_url)
+    request_func = getattr(requests, method)
+    headers.update({'Authorization': 'Bearer {}'.format(access_token)})
+    r = request_func(url, headers=headers, **kwargs)
+
+    if r.status_code == 404 or r.status_code == 403:
+        logger.warning('{} {} {} {} {}'.format(method.upper(), url, r.status_code, len(r.text), user))
+        message = 'Warning: called Terra API: /{} got status: {} with a reason: {}'.format(path, r.status_code, r.reason)
+        if r.status_code == 404:
+            raise TerraNotFoundException(message)
+        raise PermissionDenied(message)
```

Comment:
this would be a lot cleaner if you did separate if statements for 403 and 404. Also all 40X responses will be logged as warnings, so I don't think you need to also explicitly log here
```
if r.status_code == 404:
        raise TerraNotFoundException('Warning: called Terra API: {} /{} got status 404 with reason: {}'.format(method.upper(), path, r.reason)
if r.status_code == 403:
        raise PermissionDenied('Access denied (403) from Terra API: {} /{} with reason: {}'.format(method.upper(), path, r.reason)
```

Reply:
Okay

Reply:
It seems that for the Terra API. There is no other place that will process the status code. So the log is required here or the function which handles the exception.

Reply:
unhandled exceptions in seqr are handled by the middleware and logged as exceptions. If you don't want that to happen, then you need to add exception handling for these errors whenever `anvil_call` is called, or you need to update middleware.py to properly handle them. But at any rate, something will eventually receive a `TerraNotFoundException` and either it will choose to ignore that exception, in which case not logging is fine, or the exception will bubble up and be logged in the middleware. I thin the best bet is to update `EXCEPTION_ERROR_MAP` in middleware.py to map a `TerraNotFoundException` to a 404